patches/chromium/PipeWire-capturer-implement-proper-DMA-BUFs-support.patch

```diff
@@ -0,0 +1,421 @@
++    RTC_LOG(LS_ERROR) << "Error during obtaining EGL display: "
++                      << formatGLError(eglGetError());
++    return;
++  }
++
++  EGLint major, minor;
++  if (eglInitialize(egl_.display, &major, &minor) == EGL_FALSE) {
++    RTC_LOG(LS_ERROR) << "Error during eglInitialize: "
++                      << formatGLError(eglGetError());
++    return;
++  }
++
++  if (eglBindAPI(EGL_OPENGL_API) == EGL_FALSE) {
++    RTC_LOG(LS_ERROR) << "bind OpenGL API failed";
++    return;
++  }
++
++  egl_.context =
++      eglCreateContext(egl_.display, nullptr, EGL_NO_CONTEXT, nullptr);
++
++  if (egl_.context == EGL_NO_CONTEXT) {
++    RTC_LOG(LS_ERROR) << "Couldn't create EGL context: "
++                      << formatGLError(eglGetError());
++    return;
++  }
++
++  RTC_LOG(LS_ERROR) << "Egl initialization succeeded";
++
++  egl_initialized_ = true;
++}
++
+ void BaseCapturerPipeWire::InitPortal() {
+   cancellable_ = g_cancellable_new();
+   g_dbus_proxy_new_for_bus(
+@@ -697,9 +807,10 @@ void BaseCapturerPipeWire::HandleBuffer(pw_buffer* buffer) {
+     return;
+   }
+ 
++  std::function<void()> cleanup;
++  const int32_t src_stride = spaBuffer->datas[0].chunk->stride;
+ #if PW_CHECK_VERSION(0, 3, 0)
+-  if (spaBuffer->datas[0].type == SPA_DATA_MemFd ||
+-      spaBuffer->datas[0].type == SPA_DATA_DmaBuf) {
++  if (spaBuffer->datas[0].type == SPA_DATA_MemFd) {
+ #else
+   if (spaBuffer->datas[0].type == pw_core_type_->data.MemFd ||
+       spaBuffer->datas[0].type == pw_core_type_->data.DmaBuf) {
+@@ -711,7 +822,7 @@ void BaseCapturerPipeWire::HandleBuffer(pw_buffer* buffer) {
+                  PROT_READ, MAP_PRIVATE, spaBuffer->datas[0].fd, 0)),
+         spaBuffer->datas[0].maxsize + spaBuffer->datas[0].mapoffset,
+ #if PW_CHECK_VERSION(0, 3, 0)
+-        spaBuffer->datas[0].type == SPA_DATA_DmaBuf,
++        false,
+ #else
+         spaBuffer->datas[0].type == pw_core_type_->data.DmaBuf,
+ #endif
+@@ -723,15 +834,99 @@ void BaseCapturerPipeWire::HandleBuffer(pw_buffer* buffer) {
+       return;
+     }
+ 
+-#if PW_CHECK_VERSION(0, 3, 0)
+-    if (spaBuffer->datas[0].type == SPA_DATA_DmaBuf) {
+-#else
++#if !PW_CHECK_VERSION(0, 3, 0)
+     if (spaBuffer->datas[0].type == pw_core_type_->data.DmaBuf) {
+-#endif
+       SyncDmaBuf(spaBuffer->datas[0].fd, DMA_BUF_SYNC_START);
+     }
++#endif
+ 
+     src = SPA_MEMBER(map.get(), spaBuffer->datas[0].mapoffset, uint8_t);
++#if PW_CHECK_VERSION(0, 3, 0)
++  } else if (spaBuffer->datas[0].type == SPA_DATA_DmaBuf) {
++    if (!egl_initialized_) {
++      // Shouldn't reach this
++      RTC_LOG(LS_ERROR) << "Failed to process DMA buffer.";
++      return;
++    }
++
++    gbm_import_fd_data importInfo = {
++        static_cast<int>(spaBuffer->datas->fd),
++        static_cast<uint32_t>(desktop_size_.width()),
++        static_cast<uint32_t>(desktop_size_.height()),
++        static_cast<uint32_t>(spaBuffer->datas[0].chunk->stride),
++        GBM_BO_FORMAT_ARGB8888};
```

Comment:
Are we sure the buffer will have this format?

Reply:
It's either GBM_BO_FORMAT_ARGB8888 or GBM_BO_FORMAT_XRGB8888. I don't think it matters, either the alpha channel is ignored or not, it will be processed into correct format later anyway.

Reply:
Where is this ensured?